docs/concurrency-adoption-guidelines.md

```diff
@@ -0,0 +1,179 @@
+# Swift Concurrency adoption guidelines for Swift Server Libraries
+
+This writeup attempts to provide a set of guidelines to follow by authors of server-side swift libraries. Specifically a lot of the discussion here revolves around what to do about existing APIs and libraries making extensive use of Swift NIO’s EventLoopFuture and related types.
+
+Swift Concurrency is a multi-year effort. It is very valuable for the server community to participate in this multi-year adoption of the concurrency features, one by one, and provide feedback while doing so. As such, we should not hold off adopting concurrency features until Swift 6 as we may miss valuable opportunity to improve the concurrency model.
+
+In 2021 we saw structured concurrency and actors. Now is a great time to provide APIs using those primitives. In the future we will see fully checked Swift concurrency. This will come with breaking changes. For this reason adopting the new concurrency features can be split into two phases.
+
+
+## What you can do right now
+
+### API Design
+
+Firstly, existing libraries should strive to add `async` functions where possible to their user-facing “surface” APIs in addition to existing `*Future` based APIs wherever possible. These additive APIs can be gated on the Swift version and can be added without breaking existing users, for example like this:
+
+```swift
+extension Worker {
+  func work() -> EventLoopFuture<Value> { ... }
+  
+  #if swift(>=5.5)
+  func work() async throws -> Value { ... }
+  #endif
+}
+```
+
+If a function cannot fail but was using futures before, it should not include the throws keyword in its new incarnation. 
+
+Such adoption can begin immediately, and should not cause any issues to existing users of existing libraries. 
+
+### SwiftNIO helper functions
+
+To allow an easy transition to async code, SwiftNIO offers a number of helper methods on EventLoopFuture and -Promise. Those live in the `_NIOConcurrency` module and will move to NIOCore once Swift concurrency is released.
+
+On every `EventLoopFuture` you can call `.get()` to transition the future into an awaitable invocation. If you want to translate async/await calls to an EventLoopFuture we recommend the following pattern:
+
+```swift 
+#if swift(>=5.5)
+func yourAsyncFunctionConvertedToAFuture(on eventLoop: EventLoop) 
+    -> EventLoopFuture<Result> {
+    let promise = context.eventLoop.makePromise(of: Out.self)
```

Comment:
the return type of the method is `EventLoopFuture<Result>` but here we make an `EventLoopPromise<Out>`, should the generic types match?

Reply:
Oh, thanks 👍

Reply:
I think you don't need `context.` because `eventLoop` is passed directly as a function parameter.